src/compiler/parser.ts

```diff
@@ -1086,7 +1086,16 @@ namespace ts {
             return currentToken;
         }
 
-        function nextToken(): SyntaxKind {
+        function nextToken(tokenReparsedAsIdentifier?: boolean): SyntaxKind {

```

Comment:
We call `nextToken` a lot when we parse, especially in the language service, and adding an optional argument makes the function polymorphic. I would specialize this case with its own function.

Reply:
Fair enough - a function with a Boolean flag on a branch may as well just be another function.